src/main/scala/org/ergoplatform/modifiers/history/popow/NipopowAlgos.scala

```diff
@@ -0,0 +1,298 @@
+
+
+/**
+  * A set of utilities for working with NiPoPoW protocol.
+  *
+  * Based on papers:
+  *
+  * [KMZ17] Non-Interactive Proofs of Proof-of-Work, FC 20 (published) version
+  *           https://fc20.ifca.ai/preproceedings/74.pdf
+  *
+  * [KLS16] Proofs of Proofs of Work with Sublinear Complexity http://fc16.ifca.ai/bitcoin/papers/KLS16.pdf
+  *
+  * Please note that for [KMZ17] we're using the version published @ Financial Cryptography 2020, which is different
+  * from previously published versions on IACR eprint.
+  */
+class NipopowAlgos(powScheme: AutolykosPowScheme) {
+  import NipopowAlgos._
+
+  /**
+    * Computes interlinks vector for a header next to `prevHeader`.
+    */
+  @inline def updateInterlinks(prevHeaderOpt: Option[Header], prevExtensionOpt: Option[Extension]): Seq[ModifierId] =
+    prevHeaderOpt.flatMap { prevHeader =>
+      prevExtensionOpt
+        .flatMap(ext => NipopowAlgos.unpackInterlinks(ext.fields).toOption)
+        .map(updateInterlinks(prevHeader, _))
+    }.getOrElse(Seq.empty)
+
+  /**
+    * Computes interlinks vector for a header next to `prevHeader`.
+    */
+  @inline def updateInterlinks(prevHeader: Header, prevInterlinks: Seq[ModifierId]): Seq[ModifierId] =
+    if (!prevHeader.isGenesis) {
+      require(prevInterlinks.nonEmpty, "Interlinks vector could not be empty in case of non-genesis header")
+      val genesis = prevInterlinks.head
+      val tail = prevInterlinks.tail
+      val prevLevel = maxLevelOf(prevHeader)
+      if (prevLevel > 0) {
+        (genesis +: tail.dropRight(prevLevel)) ++ Seq.fill(prevLevel)(prevHeader.id)
+      } else {
+        prevInterlinks
+      }
+    } else {
+      Seq(prevHeader.id)
+    }
+
+  /**
+    * Packs interlinks into key-value format of the block extension.
+    */
+  @inline def packInterlinks(links: Seq[ModifierId]): Seq[(Array[Byte], Array[Byte])] = {
+    @scala.annotation.tailrec
+    def loop(rem: List[(ModifierId, Int)],
+             acc: Seq[(Array[Byte], Array[Byte])]): Seq[(Array[Byte], Array[Byte])] =
+      rem match {
+        case (headLink, idx) :: _ =>
+          val duplicatesQty = links.count(_ == headLink)
+          val filled = Array(InterlinksVectorPrefix, idx.toByte) -> (duplicatesQty.toByte +: idToBytes(headLink))
+          loop(rem.drop(duplicatesQty), acc :+ filled)
+        case Nil =>
+          acc
+      }
+
+    loop(links.zipWithIndex.toList, Seq.empty)
+  }
+
+  @inline def interlinksToExtension(links: Seq[ModifierId]): ExtensionCandidate =
+    ExtensionCandidate(packInterlinks(links))
+
+  /**
+    * Computes max level (μ) of the given [[Header]], such that μ = log(T) − log(id(B))
+    */
+  def maxLevelOf(header: Header): Int =
+    if (!header.isGenesis) {
+      val requiredTarget = org.ergoplatform.mining.q / RequiredDifficulty.decodeCompactBits(header.nBits)
+      val realTarget = powScheme.powHit(header).doubleValue
+      val level = log2(requiredTarget.doubleValue) - log2(realTarget.doubleValue)
```

Comment:
Can we avoid this both the log2 and bigint -> double casts here?

Reply:
Critical problem: the maximum target must be used here instead of the block target deduced from nBits